cmip6_downscaling/analysis/analysis.py

```diff
@@ -0,0 +1,183 @@
+connection_string = os.environ.get("AZURE_STORAGE_CONNECTION_STRING")
+fs = fsspec.filesystem('az', connection_string=connection_string)
+
+
+def qaqc_checks(ds):
+    # run_qaqc_dir =  f"{analysis_dir}/qaqc/{run_id}"
+    qaqc_ds = make_qaqc_ds(ds)
+    annual_qaqc_ts = qaqc_ds.groupby('time.year').sum().sum(dim=['lat', 'lon']).to_dataframe()
+    qaqc_maps = qaqc_ds.sum(dim='time')
+    return annual_qaqc_ts, qaqc_maps
+
+
+def monthly_summary(ds):
+    """Creates an monthly summary dataset.
+
+    Parameters
+    ----------
+    ds : xr.Dataset
+        Downscaled dataset
+    """
+    out_ds = xr.Dataset()
+    for var in ds:
+        if var in ['tasmax', 'tasmin']:
+            out_ds[var] = ds[var].resample(time='1MS').mean(dim='time')
+        elif var in ['pr']:
+            out_ds[var] = ds[var].resample(time='1MS').sum(dim='time')
+        else:
+            print(f'{var} not implemented')
+
+    return out_ds
+
+
+def annual_summary(ds):
+    """Creates an annual summary dataset.
+
+    Parameters
+    ----------
+    ds : xr.Dataset
+        Downscaled dataset
+    """
+    out_ds = xr.Dataset()
+    for var in ds:
+        if var in ['tasmax', 'tasmin']:
+            out_ds[var] = ds[var].resample(time='YS').mean()
+        elif var in ['pr']:
+            out_ds[var] = ds[var].resample(time='YS').sum()
+        else:
+            print(f'{var} not implemented')
+
+    return out_ds
+
+
+@task(log_stdout=True, tags=['dask-resource:TASKSLOTS=1'])
+def run_analyses(parameters, web_blob):
+    run_id = parameters['run_id']
+    workdir = '/home/jovyan/cmip6-downscaling/notebooks'
+    executed_notebook_path = f'{workdir}/analyses_{run_id}.ipynb'
+    executed_notebook_html_path = f'{workdir}/analyses_{run_id}.html'
+    pm.execute_notebook(
+        f'{workdir}/analyses.ipynb', executed_notebook_path, parameters=parameters
+    )  # TODO do these local paths work regardless of execution environment?
+
+    # convert from ipynb to html
+    os.system(f"jupyter nbconvert {executed_notebook_path} --to html")
+
+    blob_service_client = BlobServiceClient.from_connection_string(connection_string)
+    blob_name = web_blob + f'/analyses_{run_id}.html'
+    blob_client = blob_service_client.get_blob_client(container='$web', blob=blob_name)
+    # clean up before writing
+    try:
+        blob_client.delete_blob()
+    except:
+        pass
+
+    #  need to specify html content type so that it will render and not download
+    with open(executed_notebook_html_path, "rb") as data:
+        blob_client.upload_blob(data, content_settings=ContentSettings(content_type='text/html'))
+
+    return executed_notebook_path
+
+
+def load_top_cities(plot=False):
+    cities = pd.read_csv('/home/jovyan/cmip6-downscaling/notebooks/worldcities.csv')
+    top_cities = (
+        cities.sort_values('population', ascending=False)
+        .groupby('country')
+        .first()
+        .sort_values('population', ascending=False)[0:100][['city', 'lat', 'lng']]
```

Comment:
suggest adding the number of `top_cities` as a parameter in the function signature. And probably also `additional_cities` as a option/boolean flag to opt in/out.

Reply:
great! done.